janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/subquerycache/SubqueryCache.java

```diff
@@ -0,0 +1,112 @@
+ * <p>
+ * For example, suppose the cache is empty at the beginning. Firstly, prop1_idx:multiKSQ[1]@100 query leads to a cache
+ * miss, and then results are loaded into cache. Secondly, prop1_idx:multiKSQ[1]@20 query leads to a cache hit, and
+ * then first 20 results of cached responses are returned. Thirdly, prop1_idx:multiKSQ[1]@2000 leads to a cache miss,
+ * then results are loaded into cache, and initial results saved by prop1_idx:multiKSQ[1]@100 query are overriden by
+ * the new results.
+ * <p>
+ * Internally, raw query results are encapsulated in {@link SubqueryResult} object together with limit of the query.
+ * Meanwhile, keys are stored without limit (or with a dummy limit). Whenever there is a hit for key, the limit of
+ * cached results and that of the query will be compared, and whether this is a cache hit or miss will be determined.
+ * <p>
+ * Compared to a simply {@link JointIndexQuery.Subquery} to {@code List<Object>} mapping, this cache has two benefits:
+ * 1) it reduces latency, as "subset" queries can effectively leverage cached results.
+ * 2) it saves space, as queries only differing in limit will not be saved more than once in cache.
+ *
+ * @author Boxuan Li (liboxuan@connect.hku.hk)
+ */
+public interface SubqueryCache {
+
+    /**
+     * Return a list of results if given query exists in cache, otherwise
+     * return null
+     *
+     * @param query a subquery of joint index query
+     * @return a list of matching results or null if is not in the cache
+     */
+    List<Object> getIfPresent(JointIndexQuery.Subquery query);
+
+    /**
+     * Add given values into cache
+     *
+     * @param query  a subquery of joint index query
+     * @param values a list of results to be cached
+     */
+    void put(JointIndexQuery.Subquery query, List<Object> values);
+
+    /**
+     * Returns a list of results. If given query not in cache, call the
+     * value loader to retrieve results and put into the cache
+     *
+     * @param query       a subquery of joint index query
+     * @param valueLoader a callable that returns a list of results
+     * @return a list of results
+     * @throws Exception if exception thrown when calling the value loader
+     */
+    List<Object> get(JointIndexQuery.Subquery query, Callable<? extends List<Object>> valueLoader) throws Exception;
+
+    /**
+     * Closes the cache which allows the cache to release allocated memory.
+     * Calling any of the other methods after closing a cache has undetermined behavior.
+     */
+    void close();
+
+    public class SubqueryResult {
+        private List<Object> values;
+        private int limit;
+
+        protected SubqueryResult(List<Object> values, int limit) {
+            this.values = values;
+            this.limit = limit;
+        }
+
+        public List<Object> getValues() {
+            return values;
+        }
+
+        public void setValues(List<Object> values) {
```

Comment:
Nitpick: Do `values` and `limit` have to be modifiable once their values are set by the constructor?

Reply:
make sense, will remove redundant setters